stellargraph/layer/knowledge_graph.py

```diff
@@ -191,3 +192,135 @@ def build(self):
         x_out = self(x_inp)
 
         return x_inp, x_out
+
+
+class DistMultScore(Layer):
+    """
+    DistMult scoring Keras layer.
+
+    Original Paper: Embedding Entities and Relations for Learning and Inference in Knowledge
+    Bases. Bishan Yang, Wen-tau Yih, Xiaodong He, Jianfeng Gao, Li Deng. ICLR 2015
+
+    This combines subject, relation and object embeddings into a score of the likelihood of the
+    link.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shape):
+        self.built = True
+
+    def call(self, inputs):
+        e1, r, e2 = inputs
```

Comment:
I mention this below as well, but I think these should be called y_e1, y_e2, and m_r (or M_r). The notation in the paper is a bit confusing because in equation 3, they use the (general) notation S_(e1, r, e2) (note the underscore) to refer to equation 2 where it actually translates to g_r^b(y_e1, y_e2). So the input are the node embeddings not the one-hot node values.

Reply:
You're right that that's closer to the paper. I'd elided the `y`/`M` notation because it seemed context would be fine (and because `M_r` itself doesn't actually exist anywhere), but now that you pointed it out, I definitely agree that it's clearer to match the paper.

Also, minor detail, but there's not strictly any one-hot node values in this code, only sequential node identifiers in the form of "ilocs" (but, of course, they're equivalent).